pkg/workspace/controllers/garbagecollect/pv_controller.go

```diff
@@ -0,0 +1,173 @@
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/component-helpers/storage/volume"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/kaito-project/kaito/pkg/utils/consts"
+)
+
+type PersistentVolumeGCReconciler struct {
+	client.Client
+	Log      logr.Logger
+	Scheme   *runtime.Scheme
+	Recorder record.EventRecorder
+}
+
+func NewPersistentVolumeGCReconciler(client client.Client, scheme *runtime.Scheme, log logr.Logger, Recorder record.EventRecorder) *PersistentVolumeGCReconciler {
+	return &PersistentVolumeGCReconciler{
+		Client:   client,
+		Scheme:   scheme,
+		Log:      log,
+		Recorder: Recorder,
+	}
+}
+
+func (c *PersistentVolumeGCReconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	pvObj := &corev1.PersistentVolume{}
+	if err := c.Client.Get(ctx, req.NamespacedName, pvObj); err != nil {
+		klog.ErrorS(err, "Failed to get PersistentVolume", "name", req.Name, "namespace", req.Namespace)
+		return reconcile.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if pvObj.Spec.ClaimRef == nil {
+		// No claim reference, nothing to do
+		return reconcile.Result{}, nil
+	}
+
+	pvcNamespacedName := types.NamespacedName{
+		Namespace: pvObj.Spec.ClaimRef.Namespace,
+		Name:      pvObj.Spec.ClaimRef.Name,
+	}
+
+	pvcObj := &corev1.PersistentVolumeClaim{}
+	if err := c.Client.Get(ctx, pvcNamespacedName, pvcObj); err != nil {
+		if client.IgnoreNotFound(err) == nil {
+			klog.InfoS("PVC not found, force deleting PV", "pv", klog.KObj(pvObj), "pvc", pvcNamespacedName)
+			return reconcile.Result{}, forceDeletePV(ctx, c.Client, pvObj)
+		}
+		klog.ErrorS(err, "Failed to get PersistentVolumeClaim", "pvc", pvcNamespacedName)
+		return reconcile.Result{}, err
+	}
+
+	result, err := c.deleteOrphanedLocalPV(ctx, pvObj, pvcObj)
+	if err != nil {
+		klog.ErrorS(err, "Failed to delete orphaned local PV", "pv", klog.KObj(pvObj), "pvc", klog.KObj(pvcObj))
+		c.Recorder.Eventf(pvcObj, corev1.EventTypeWarning, "DeletionFailed", "Failed to delete orphaned local PV %s: %v", klog.KObj(pvObj), err)
+		return result, err
+	}
+
+	return result, nil
+}
+
+// When a node with local volumes gets removed from a cluster before deleting those volumes,
+// the PV and PVC objects may still exist. Force deleting objects in this case.
+// Ref: https://github.com/kubernetes-csi/external-provisioner?tab=readme-ov-file#deleting-local-volumes-after-a-node-failure-or-removal
+func (c *PersistentVolumeGCReconciler) deleteOrphanedLocalPV(ctx context.Context, pv *corev1.PersistentVolume, pvc *corev1.PersistentVolumeClaim) (ctrl.Result, error) {
```

Comment:
If we use `deleteOrphanedLocalPV` as function name, we need to move `check node exist` routine out of this function. In this function, we will delete orphaned local pv only.